### PR TITLE
[BugFix] Fix the disorder when enable local_pass_through

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：

Things have been a little weird while pass_through is enabled. Consider such a situation:
1. time1: ChunkA is added to pass_through context, now context has only ChunkA. And related RequestA(sequence=1) is sent.
1. time2: ChunkB is added to pass_through context, now context has ChunkA and ChunkB. And related RequestB(sequence=2) is sent.
1. time3: RequestB is received, and all the chunks(ChunkA and ChunkB) have been fetched, and they are considered belonging to RequestB.
1. time4: ChunkC is added to pass_through context, now context has only ChunkC. And related RequestC(sequence=3) is sent.
1. tiem5: RequestA is received, and all the chunks(ChunkC) have been fetched, and they are considered belonging to RequestA.
1. time6: RequestC is received, nothing can be fetched.

And here comes the funny thing, RequestA received the ChunkC, and RequestB received the ChunkA and ChunkB, so with the help of the keep_order mechanism, the output order will be ChunkC, ChunkA, ChunkB. But the expected order is ChunkA, ChunkB, ChunkC.

So the solution is simple, if pass_through is enabled, simply treat it as un-ordered request, and the order will be kept naturally.

## Why this disorder does not result in wrong results?

Currently, Merge Exchange use a K-merge algorighm to gather all streams from all senders, and the algorithm itself can tolerant a relatively little disorder(basically, it's a heap).

And recently, I'm working on another parallel merge algorithm(https://github.com/StarRocks/starrocks/issues/20534), which requires the data stream from the same sender must be exactly ordered.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/runtime/sender_queue.cpp [hunk 1]
```diff
@@ -730,17 +752,23 @@ Status DataStreamRecvr::PipelineSenderQueue::add_chunks(const PTransmitChunkPara
...
             *done = nullptr;
         }
+
         for (auto& chunk : chunks) {
             int index = _is_pipeline_level_shuffle ? chunk.driver_sequence : 0;
             size_t chunk_bytes = chunk.chunk_bytes;
             auto* closure = chunk.closure;
-            _chunk_queues[index].enqueue(std::move(chunk));
+            if (keep_order && use_pass_through) {
+                _chunk_queues[index].enqueue(*_producer_token, std::move(chunk));
```
Comment (dirtysalt) [inline be/src/runtime/sender_queue.cpp:768]: I don't get why here we should have `producer_token`? to keep order?

**Review by dirtysalt — DISMISSED** [review]

File: be/src/runtime/sender_queue.cpp [hunk 2]
```diff
@@ -646,18 +646,32 @@ Status DataStreamRecvr::PipelineSenderQueue::add_chunks(const PTransmitChunkPara
...
-    if (keep_order) {
+    // Things have been a little weird while pass_through is enabled
+    // Consider such a situation:
+    //      time1: ChunkA is added to pass_through context, now context has only ChunkA. And related RequestA(sequence=1) is sent.
+    //      time2: ChunkB is added to pass_through context, now context has ChunkA and ChunkB. And related RequestB(sequence=2) is sent.
+    //      time3: RequestB is received, and all the chunks(ChunkA and ChunkB) have been fetched, and they are considered belonging to RequestB.
+    //      time4: ChunkC is added to pass_through context, now context has only ChunkC. And related RequestC(sequence=3) is sent.
+    //      tiem5: RequestA is received, and all the chunks(ChunkC) have been fetched, and they are considered belonging to RequestA.
+    //      time6: RequestC is received, nothing can be fetched.
+    // And here comes the funny thing, RequestA received the ChunkC, and RequestB received the ChunkA and ChunkB, so with the help of the
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply be/src/runtime/sender_queue.cpp:670]: No, with. The keep_order machanism is not compatible with the local_pass_through scenerio.

File: be/src/runtime/sender_queue.cpp (hunk 1 above)
Comment (liuyehcf) [reply be/src/runtime/sender_queue.cpp:768]: The chunk enqueue first must be dequeue first either.

If withouth this producer_token, for instance, chunk1 and chunk2 are enqueued in order, but may be chunk2 is dequeue first.

**Review by dirtysalt — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by liuyehcf** [comment]
@Mergifyio backport branch-3.0 branch-2.5 branch-2.4 branch-2.3

**Comment by liuyehcf** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**